src/Nethermind/Nethermind.Optimism/CL/P2P/PayloadDecoder.cs

```diff
@@ -0,0 +1,114 @@
+// SPDX-FileCopyrightText: 2024 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using Nethermind.Core;
+using Nethermind.Merge.Plugin.Data;
+
+namespace Nethermind.Optimism.CL;
+
+public class PayloadDecoder : IPayloadDecoder
+{
+    public static readonly PayloadDecoder Instance = new();
+
+    private const int PrefixDataSize = 560;
+
+    private PayloadDecoder()
+    {
+    }
+
+    public ExecutionPayloadV3 DecodePayload(ReadOnlySpan<byte> data)
+    {
+        ExecutionPayloadV3 payload = new();
+
+        if (PrefixDataSize >= data.Length)
+        {
+            throw new ArgumentException("Invalid payload data size");
+        }
+
+        int offset = 0;
+        payload.ParentBeaconBlockRoot = new(data[offset..(offset + 32)]);
+        offset += 32;
+        payload.ParentHash = new(data[offset..(offset + 32)]);
+        offset += 32;
+        payload.FeeRecipient = new(data[offset..(offset + 20)]);
+        offset += 20;
+        payload.StateRoot = new(data[offset..(offset + 32)]);
+        offset += 32;
+        payload.ReceiptsRoot = new(data[offset..(offset + 32)]);
+        offset += 32;
+        payload.LogsBloom = new(data[offset..(offset + 256)]);
+        offset += 256;
+        payload.PrevRandao = new(data[offset..(offset + 32)]);
+        offset += 32;
+        payload.BlockNumber = (long)BitConverter.ToUInt64(data[offset..(offset + 8)]);
```

Comment:
Not sure about this one but I got suggested some time ago that we should use `BinaryPrimitives` rather than `BitConverter` (performance?)